### ISSUE TITLE
flash fails to start with Galaxy Note 3 SM-N9005

### ISSUE THREAD

**Issue by dpocock** [Benjamin-Dobell/Heimdall#197]
I've observed the same issue that is described in this thread: http://forum.xda-developers.com/showthread.php?t=2684410

I run

    heimdall flash --recovery something.tar

and it fails:

    Downloading device's PIT file...
    ERROR: Failed to send request to end PIT file transfer!
    ERROR: Failed to download PIT file!
    Ending session...
    ERROR: Failed to send end session packet!
    Releasing device interface...

**Comment by codyharrington** [comment]
Just wanting to confirm that @karabijavad's fork worked for me.

I cloned it and compiled it from source, and it ran flawlessly.

Samsung Galaxy S4 LTE+ (GT-i9506)
Running on Arch Linx x64
Output of uname -a:
`Linux citadel 3.17.6-1-ARCH #1 SMP PREEMPT Sun Dec 7 23:43:32 UTC 2014 x86_64 GNU/Linux`

@Benjamin-Dobell could you incorporate this fix into the main branch perhaps? Or is it too risky?

**Comment by davemc50** [comment]
@karabijavad's fork worked on first run only, successfully gave print-pit output, and did not choke on protocol initialisation.  Subsequent runs (even after os reboot) give the failed handshake -7.

Ubuntu 64bit 14.04; Samsung Galaxy S II i9100;
 libusb-1.0-0-dev:amd64                      2:1.0.17-1ubuntu2         

Could someone add some more detailed debug code in this area onwards....
[00000f84] libusbx: debug [handle_events] poll() returned 1
[00000f84] libusbx: debug [handle_events] timerfd triggered
[00000f84] libusbx: debug [libusb_cancel_transfer]

**Comment by yarolig** [comment]
I used @karabijavad branch add it works for me.
Runned these commands.
```
 git clone https://github.com/karabijavad/Heimdall.git 
 cd Heimdall/
 git checkout 02ccbbf
 cd libpit/
 make
 ./configure
 make
 cd ..
 cd heimdall
 ./configure
 make -j 4
 sudo make install
 sudo udevadm control --reload-rules
 cd to/recoveries
 sudo heimdall flash --RECOVERY openrecovery-twrp-2.7.1.0-hlte-4.4.img
```
flash-command works well.
download-pid and print-pit modes works once. Second command will fail unless I unplug cable and plug it again.
```
$ sudo heimdall flash --RECOVERY openrecovery-twrp-2.7.1.0-hlte-4.4.img
Heimdall v1.4.1

Copyright (c) 2010-2014 Benjamin Dobell, Glass Echidna
http://www.glassechidna.com.au/

This software is provided free of charge. Copying and redistribution is
encouraged.

If you appreciate this software and you would like to support future
development please consider donating:
http://www.glassechidna.com.au/donate/

Initialising connection...
Detecting device...
Claiming interface...
Attempt failed. Detaching driver...
Claiming interface again...
Setting up interface...

Initialising protocol...
Protocol initialisation successful.

Beginning session...

Some devices may take up to 2 minutes to respond.
Please be patient!

Session begun.

Downloading device's PIT file...
PIT file data received, read 'empty' packet...
PIT file data receive completed, send kRequestEndTransfer...
PIT file download successful.

Uploading RECOVERY
100%
RECOVERY upload successful

Ending session...
Rebooting device...
Releasing device interface...
Re-attaching kernel driver...

$ 
```
By the way.
For some unknown reason CWM-recovery can't restore my backups.
I recommend to make a backup with TWRP too.

**Comment by tzewangdorje** [comment]
Just wanted to mention, in case it helps anyone else. When I first ran the version compiled from source from 02ccbbf (Yarolig's instructions were spot on here https://github.com/Benjamin-Dobell/Heimdall/issues/197#issuecomment-71892522) I too for the usb -7 error reported by some above. What worked for me was trying a different usb port - might be worth a try if you get this error.

**Comment by wkurosawa** [comment]
Thanks @yarolig ! Followed your instructions on using @karabijavad branch, and it worked flawlessly!

**Comment by nnscr** [comment]
Thanks for the tip! @karabijavad fork fixes the bug! :clap:
I'm using a non-branded Note 3 (SM-N9005).

**Comment by fferraribr** [comment]
Guys... I didn't figured out how to compile @karabijavad branch on msys2. I'm kind of noob on that. Could please someone post step by step like cyanogen wiki below? Really appreciate it!

$ pacman -Syu
$ pacman -S mingw-w64-x86_64 mingw-w64-x86_64-clang mingw-w64-x86_64-cmake mingw-w64-x86_64-libusb mingw-w64-x86_64-qt5-static make git
$ export PATH="/mingw64/bin:$PATH"
$ git clone git://github.com/Benjamin-Dobell/Heimdall.git
$ mkdir -p Heimdall/build
$ cd Heimdall/build
$ cmake -G "MSYS Makefiles" -DCMAKE_BUILD_TYPE=Release -DQt5Widgets_DIR=/mingw64/qt5-static/lib/cmake/Qt5Widgets ..
$ make

**Comment by arrSea** [comment]
In case it helps anyone, I resolved the "Result; -7" error by experimenting with different USB ports and unplugging any unnecessary USB items.

I am running Ubuntu 14.04 and heimdall v1.4.1

More details in my first comment on https://github.com/Benjamin-Dobell/Heimdall/issues/230

**Comment by Benjamin-Dobell** [comment]
Has anyone who previously experienced this issue tried using a version of master that was built after February 21st? As far as I can tell this issue should already be resolved.

If you're on OS X you can try this package (which I compiled just an hour or so ago):
https://bitbucket.org/benjamin_dobell/heimdall/downloads/Heimdall-1.4.1-Unofficial-Signed.dmg

**Comment by arrSea** [comment]
I'm not sure how relevant my particular case was, but for what it's worth, I am using a clean checkout of master.

The latest commit is:  d0526a3b74a003dfc6f805682693be9173ffcd88
                       Date:   Sat Mar 21 14:53:43 2015 +1100

As I stated in my previous comment, it worked perfectly once I unhooked unnecessary USB devices and found the magic USB port.  Thank you very much for this extremely useful tool!
